src/operator/nn/mkldnn/mkldnn_sum.cc

```diff
@@ -49,23 +49,42 @@ void Sum(const mkldnn::memory &arr1, const mkldnn::memory &arr2,
 void MKLDNNSumForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
                       const std::vector<NDArray> &inputs, const OpReqType &req,
                       const NDArray &out_data) {
+  if (req == kNullOp) {
+    return;
+  }
+
   TmpMemMgr::Get()->Init(ctx.requested[0]);
   std::vector<mkldnn::primitive::at> in_prims;
   std::vector<mkldnn::memory::primitive_desc> in_pds(inputs.size());
   std::vector<float> scales(inputs.size(), 1);
   in_prims.reserve(inputs.size());
+  bool pd_same = true;
   for (size_t i = 0; i < inputs.size(); i++) {
     auto in_mem = inputs[i].GetMKLDNNData();
     in_prims.push_back(*in_mem);
     in_pds[i] = in_mem->get_primitive_desc();
+    pd_same = pd_same && (in_pds[i] == in_pds[0]);
   }
-  mkldnn::sum::primitive_desc pdesc(scales, in_pds);
 
-  auto out_mem = CreateMKLDNNMem(out_data, pdesc.dst_primitive_desc(), req);
-  MKLDNNStream *stream = MKLDNNStream::Get();
-  stream->RegisterPrim(mkldnn::sum(pdesc, in_prims, *out_mem.second));
-  CommitOutput(out_data, out_mem);
-  stream->Submit();
+  mkldnn::sum::primitive_desc pdesc(scales, in_pds);
+  pd_same = pd_same && (pdesc.dst_primitive_desc() == in_pds[0]);
+  auto out_mem = const_cast<NDArray&>(out_data).CreateMKLDNNData(pdesc.dst_primitive_desc());
+  bool addr_same = out_mem->get_data_handle() == inputs[0].GetMKLDNNData()->get_data_handle();
```

Comment:
You already have mkldnn memory for `inputs[0]`. you don't need to call it again to get it.

Reply:
Is GetMKLDNNData a heavy operation here? I think it will return quickly if mkl_mem_ has already been created. If I try to get memory address from in_prims[0], I still need cast it to type mkldnn::memory, since it is with type mkldnn::primitive::at.

Reply:
my PR tries to make GetMKLDNNData() lightweight. OK, I guess it's not a big problem.